libcob/fileio.c

```diff
@@ -920,19 +920,29 @@ bdb_bt_compare (DB *db, const DBT *k1, const DBT *k2
 {
 	const unsigned char *col = (unsigned char *)DBT_GET_APP_DATA (k1);
 	COB_UNUSED (db);
-
+#ifdef USE_BDB_KEYDIFF /* flag passed with CPPFLAGS */
+	COB_MODULE_PTR->collating_sequence = col;
+	unsigned char *data1 = k1->data;
+	unsigned char *data2 = k2->data;
+	size_t size1 = (size_t)k1->size;
+	size_t size2 = (size_t)k2->size;
+	return cob_cmp_strings (data1, data2, size1, size2, col);
+#else
 	/* LCOV_EXCL_START */
 	if (col == NULL) {
 		cob_runtime_error ("bdb_bt_compare was set but no collating sequence was stored in DBT");
+		cob_hard_failure ();
 	}
 	if (k1->size != k2->size) {
 		cob_runtime_error ("bdb_bt_compare was given keys of different length");
+		cob_hard_failure ();
 	}
 	/* LCOV_EXCL_STOP */
 #if DB_VERSION_MAJOR >= 6
 	locp = NULL;	/* docs: must be set to NULL or corruption can occur ... */
 #endif
 	return indexed_key_compare (k1->data, k2->data, k2->size, col);
+#endif
```

Comment:
That's a longer ifdef with nesting - I suggest (not require) to add `/* USE_BDB_KEYDIFF */` after that.